Create a BlockCacheLookupContext to enable fine-grained block cache tracing.

HaoyuHuang opened the pull request:
BlockCacheLookupContext only contains the caller for now.
We will trace block accesses at five places:
1. BlockBasedTable::GetFilter.
2. BlockBasedTable::GetUncompressedDict.
3. BlockBasedTable::MaybeReadAndLoadToCache. (To trace access on data, index, and range deletion block.)
4. BlockBasedTable::Get. (To trace the referenced key and whether the referenced key exists in a fetched data block.)
5. BlockBasedTable::MultiGet. (To trace the referenced key and whether the referenced key exists in a fetched data block.)

We create the context at:
1. BlockBasedTable::Get. (kUserGet)
2. BlockBasedTable::MultiGet. (kUserMGet)
3. BlockBasedTable::NewIterator. (either kUserIterator, kCompaction, or external SST ingestion calls this function.)
4. BlockBasedTable::Open. (kPrefetch)
5. Index/Filter::CacheDependencies. (kPrefetch)
6. BlockBasedTable::ApproximateOffsetOf. (kCompaction or kUserApproximateSize).

I loaded 1 million key-value pairs into the database and ran the readrandom benchmark with a single thread. I gave the block cache 10 GB to make sure all reads hit the block cache after warmup. The throughput is comparable.
Throughput of this PR: 231334 ops/s.
Throughput of the master branch: 238428 ops/s.

Experiment setup: 
RocksDB:    version 6.2
Date:       Mon Jun 10 10:42:51 2019
CPU:        24 * Intel Core Processor (Skylake)
CPUCache:   16384 KB
Keys:       20 bytes each
Values:     100 bytes each (100 bytes after compression)
Entries:    1000000
Prefix:    20 bytes
Keys per prefix:    0
RawSize:    114.4 MB (estimated)
FileSize:   114.4 MB (estimated)
Write rate: 0 bytes/second
Read rate: 0 ops/second
Compression: NoCompression
Compression sampling rate: 0
Memtablerep: skip_list
Perf Level: 1

Load command: ./db_bench --benchmarks="fillseq" --key_size=20 --prefix_size=20 --keys_per_prefix=0 --value_size=100 --statistics --cache_index_and_filter_blocks --cache_size=10737418240 --disable_auto_compactions=1 --disable_wal=1 --compression_type=none --min_level_to_compress=-1 --compression_ratio=1 --num=1000000

Run command: ./db_bench --benchmarks="readrandom,stats" --use_existing_db --threads=1 --duration=120 --key_size=20 --prefix_size=20 --keys_per_prefix=0 --value_size=100 --statistics --cache_index_and_filter_blocks --cache_size=10737418240 --disable_auto_compactions=1 --disable_wal=1 --compression_type=none --min_level_to_compress=-1 --compression_ratio=1 --num=1000000 --duration=120

TODOs:
1. Create a caller for external SST file ingestion and differentiate the callers for iterator.
2. Integrate tracer to trace block cache accesses.

riversand963 commented on db/version_set.h:
```diff
@@ -982,7 +982,8 @@ class VersionSet {
   uint64_t ApproximateSize(Version* v, const Slice& start, const Slice& end,
-                           int start_level = 0, int end_level = -1);
+                           int start_level = 0, int end_level = -1,
+                           bool for_compaction = false);
```
Not a big fan of default value for function arguments.

HaoyuHuang replied:
I removed the default values.

riversand963 commented on db/version_set.h:
```diff
@@ -1032,10 +1033,11 @@ class VersionSet {
   uint64_t ApproximateSizeLevel0(Version* v, const LevelFilesBrief& files_brief,
-                                 const Slice& start, const Slice& end);
+                                 const Slice& start, const Slice& end,
+                                 bool for_compaction = false);
```
Can we do without default value of arguments?

HaoyuHuang replied:
I removed the default values.

riversand963 commented on db/version_set.h:
```diff
@@ -1032,10 +1033,11 @@ class VersionSet {
 
   uint64_t ApproximateSize(Version* v, const FdWithKeyRange& f,
-                           const Slice& key);
+                           const Slice& key, bool for_compaction = false);
```
ditto

HaoyuHuang replied:
I removed the default values.

HaoyuHuang commented:
1. I moved BlockCacheLookupContext to be after GetContext in functions where there is a GetContext. Otherwise, I moved it to be the last function argument. 
2. I removed the default values of functions arguments for the functions that this PR changes.

riversand963 commented on table/block_based/block_based_filter_block_test.cc:
```diff
@@ -57,8 +57,12 @@ TEST_F(FilterBlockTest, EmptyBuilder) {
                                      std::move(block), nullptr);
-  ASSERT_TRUE(reader.KeyMayMatch("foo", nullptr, uint64_t{0}));
-  ASSERT_TRUE(reader.KeyMayMatch("foo", nullptr, 100000));
+  ASSERT_TRUE(reader.KeyMayMatch(
```
Nice

HaoyuHuang replied:
Thanks. :)

riversand963 commented on table/block_based/partitioned_filter_block.h:
```diff
@@ -80,21 +80,24 @@ class PartitionedFilterBlockReader : public FilterBlockReader {
-      const Slice& prefix, const SliceTransform* prefix_extractor,
-      uint64_t block_offset = kNotValid, const bool no_io = false,
-      const Slice* const const_ikey_ptr = nullptr) override;
+  virtual bool KeyMayMatch(const Slice& key,
```
Keep `override`, remove `virtual`.

HaoyuHuang replied:
I removed virtual in places where applicable.

riversand963 commented on table/block_based/block_based_table_reader.cc:
```diff
@@ -1840,18 +1861,23 @@ FilterBlockReader* BlockBasedTable::ReadFilter(
 }
 
 CachableEntry<FilterBlockReader> BlockBasedTable::GetFilter(
+
```
Should this empty line be removed?

HaoyuHuang replied:
Removed.

riversand963 commented on table/block_based/block_based_table_reader.cc:
```diff
@@ -1840,18 +1861,23 @@ FilterBlockReader* BlockBasedTable::ReadFilter(
 }
 
 CachableEntry<FilterBlockReader> BlockBasedTable::GetFilter(
+
```
Should this empty line be removed?

HaoyuHuang replied:
Removed.

riversand963 commented on table/block_based/block_based_table_reader.cc:
```diff
@@ -2156,10 +2187,11 @@ Status BlockBasedTable::MaybeReadBlockAndLoadToCache(
 }
 
 Status BlockBasedTable::RetrieveBlock(
+
```
Empty line

HaoyuHuang replied:
Removed.

riversand963 commented on table/block_based/block_based_table_reader.cc:
```diff
@@ -3195,9 +3260,11 @@ BlockBasedTableOptions::IndexType BlockBasedTable::UpdateIndexType() {
 //  4. internal_comparator
 //  5. index_type
 Status BlockBasedTable::CreateIndexReader(
+
```
Empty line?

HaoyuHuang replied:
Removed.

riversand963 commented:
Can you run some in-memory benchmark comparing the performance with that of master, and make sure there is no regression.

HaoyuHuang commented:
Sure. I can do that. I will give an update here on Monday.

HaoyuHuang commented:
I loaded 1 million key-value pairs into the database and ran the readrandom benchmark. I gave the block cache 1 GB to make sure all reads hit the block cache after warmup. The throughput is comparable. 
Throughput of this PR: 231334 ops/s. 
Throughput of the master branch: 238428 ops/s. 

Load command: ./db_bench --benchmarks="fillseq" --key_size=20 --prefix_size=20 --keys_per_prefix=0 --value_size=100 --statistics --cache_index_and_filter_blocks --cache_size=10737418240 --disable_auto_compactions=1 --disable_wal=1 --compression_type=none --min_level_to_compress=-1 --compression_ratio=1 --num=1000000

Run command: ./db_bench --benchmarks="readrandom,stats" --use_existing_db --threads=1 --duration=120 --key_size=20 --prefix_size=20 --keys_per_prefix=0 --value_size=100 --statistics --cache_index_and_filter_blocks --cache_size=10737418240 --disable_auto_compactions=1 --disable_wal=1 --compression_type=none --min_level_to_compress=-1 --compression_ratio=1 --num=1000000

riversand963 commented on table/block_based/block_based_table_reader.cc:
```diff
@@ -3052,6 +3110,8 @@ Status BlockBasedTable::Prefetch(const Slice* const begin,
 }
 
 Status BlockBasedTable::VerifyChecksum() {
+  // TODO(haoyu): This function is called by external sst ingestion and user.
```
External sst ingestion is also called by user.

HaoyuHuang replied:
I updated the comments.

HaoyuHuang commented on table/block_based/block_based_table_reader.h:
```diff
@@ -380,14 +391,17 @@ class BlockBasedTable : public TableReader {
       FilePrefetchBuffer* prefetch_buffer,
       std::unique_ptr<const BlockContents>* compression_dict_block) const;
   Status PrefetchIndexAndFilterBlocks(
+
```
My bad. Removed.

HaoyuHuang commented on table/block_based/block_based_table_reader.cc:
```diff
@@ -1360,9 +1376,11 @@ Status BlockBasedTable::ReadCompressionDictBlock(
 }
 
 Status BlockBasedTable::PrefetchIndexAndFilterBlocks(
+
```
Removed.
